internal/controller/mariadb_controller_update.go

```diff
@@ -72,6 +72,12 @@ func (r *MariaDBReconciler) reconcileUpdates(ctx context.Context, mdb *mariadbv1
 		return ctrl.Result{Requeue: true}, nil
 	}
 
+	if result, err := r.waitForConfiguredReplication(mdb, logger); !result.IsZero() || err != nil {
+		// the error is not returned to avoid immediate requeuing
+		// and to allow the Replication phase to configure the recreated replicas
+		return result, nil
```

Comment:
You are totally right, it would be great to handle everything from `waitForConfiguredReplication`. The reason I did it like this is because if I return an error here, the loop will be terminated and that's what we need to avoid
https://github.com/mariadb-operator/mariadb-operator/blob/5c4e09f1f7523a5a7a7a3a7887122f601ed6a857/internal/controller/mariadb_controller.go#L223

Another option I was thinking about is introducing a custom error variable for the case when replication is not configured and comparing the returned error in the loop just like we already do with `apierrors.IsNotFound`. If they match, we don't break the loop but I wasn't sure if this is an elegant way. Here's the code I'm referring to:
https://github.com/mariadb-operator/mariadb-operator/blob/5c4e09f1f7523a5a7a7a3a7887122f601ed6a857/internal/controller/mariadb_controller.go#L205-L207

I implemented it in my latest commit and it works. What do you think?

Reply:
Thanks, the approach looks legit to me. I will follow up with a PR with some minor changes to better accomodate your changes in the codebase.